common/chat-parser-xml-toolcall.cpp

```diff
@@ -0,0 +1,710 @@
+                                gbnf_format_literal(key_val_sep) + " ";
+                        if (value.contains("type") && value["type"].is_string() && value["type"] == "string") {
+                            param_rules +=
+                                    "( string-arg-val | " +
+                                    builder.add_schema(name_safe + "-arg-" + encode_to_safe(key), value) + " ) ";
+                        } else {
+                            param_rules +=
+                                    builder.add_schema(name_safe + "-arg-" + encode_to_safe(key), value) + " ";
+                        }
+                        param_rules += gbnf_format_literal(form.val_end) + " ";
+                        if (!required) param_rules += ")? ";
+                    }
+                }
+
+                std::string quoted_name = name;
+                if (form.tool_start.back() == '"' && form.tool_sep[0] == '"') {
+                    quoted_name = gbnf_format_literal(name);
+                    quoted_name = quoted_name.substr(1, quoted_name.size() - 2);
+                }
+                tool_rules.push_back(builder.add_rule(name_safe + "-call",
+                        gbnf_format_literal(form.tool_start) + " " +
+                        gbnf_format_literal(quoted_name) + " " +
+                        gbnf_format_literal(form.tool_sep) + " " +
+                        param_rules + " " +
+                        gbnf_format_literal(form.tool_end)
+                ));
+            }
+            builder.add_rule("string-arg-val", make_gbnf_excluding({form.val_end}));
+            builder.add_rule("root", gbnf_format_literal(form.scope_start) + " ( " + string_join(tool_rules, " | ") + " ) " + gbnf_format_literal(form.scope_end));
+        });
+
+        // grammar trigger for tool call
+        data.grammar_lazy = true;
+        data.grammar_triggers.push_back({ COMMON_GRAMMAR_TRIGGER_TYPE_WORD, form.scope_start + form.tool_start });
+    }
+}
+
+/**
+ * Parse XML-Style tool call for given xml_tool_call_format. Return false for invalid syntax and get the position untouched.
+ * Throws xml_toolcall_syntax_exception if there is invalid syntax and cannot recover the original status for common_chat_msg_parser.
+ * form.scope_start, form.tool_sep and form.scope_end can be empty.
+ */
+inline bool parse_xml_tool_calls(common_chat_msg_parser & builder, const struct xml_tool_call_format & form) {
+    GGML_ASSERT(!form.tool_start.empty());
+    GGML_ASSERT(!form.key_start.empty());
+    GGML_ASSERT(!form.key_val_sep.empty());
+    GGML_ASSERT(!form.val_end.empty());
+    GGML_ASSERT(!form.tool_end.empty());
+
+    constexpr auto all_space = [] (auto &str) {
+        return std::all_of(str.begin(), str.end(), [](unsigned char ch) { return std::isspace(ch); });
+    };
+    // Helper to choose return false or throw error
+    constexpr auto return_error = [](common_chat_msg_parser & builder, auto &start_pos, const bool &recovery) {
```

Comment:
Is this really a `constexpr`? Same question goes for most uses scattered in here.